cpp/arcticdb/util/reliable_storage_lock.tpp

```diff
@@ -0,0 +1,242 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#include <arcticdb/util/reliable_storage_lock.hpp>
+
+#include <arcticdb/storage/store.hpp>
+#include <arcticdb/entity/types.hpp>
+#include <arcticdb/util/constants.hpp>
+#include <arcticdb/util/configs_map.hpp>
+
+#include <mutex>
+#include <folly/experimental/FunctionScheduler.h>
+
+namespace arcticdb {
+
+namespace lock {
+
+const auto SEPARATOR = '*';
+const auto EXTENDS_PER_TIMEOUT = 5u;
+
+inline StreamDescriptor lock_stream_descriptor(const StreamId &stream_id) {
+    return StreamDescriptor{stream_descriptor(
+            stream_id,
+            stream::RowCountIndex(),
+            {scalar_field(DataType::INT64, "expiration")})};
+}
+
+inline SegmentInMemory lock_segment(const StreamId &name, timestamp expiration) {
+    SegmentInMemory output{lock_stream_descriptor(name)};
+    output.set_scalar(0, expiration);
+    output.end_row();
+    return output;
+}
+
+template <class ClockType>
+ReliableStorageLock<ClockType>::ReliableStorageLock(const std::string &base_name, const std::shared_ptr<Store> store, timestamp timeout) :
+    base_name_(base_name), store_(store), timeout_(timeout) {
+    storage::check<ErrorCode::E_UNSUPPORTED_ATOMIC_OPERATION>(store_->supports_atomic_writes(), "Storage does not support atomic writes, so we can't create a lock");
+    auto s3_timeout = ConfigsMap::instance()->get_int("S3Storage.RequestTimeoutMs", 200000) * ONE_MILLISECOND;
+    if (2 * s3_timeout > timeout) {
+        log::lock().warn(
+                "Reliable Lock is created with a timeout < twice the s3 timeout. This is not recommended, as it increases the risk for a faulty lock timeout in s3."
+                "Lock timeout: {}, S3 timeout: {}", timeout, s3_timeout);
+    }
+}
+
+template <class ClockType>
+timestamp ReliableStorageLock<ClockType>::timeout() const {
+    return timeout_;
+}
+
+inline Epoch get_next_epoch(std::optional<Epoch> maybe_prev) {
+    if (maybe_prev.has_value()) {
+        return maybe_prev.value() + 1;
+    }
+    return 0;
+}
+
+template <class ClockType>
+StreamId ReliableStorageLock<ClockType>::get_stream_id(Epoch e) const {
```

Comment:
e is commonly used for error, no harm in calling it 'epoch'

Reply:
Used `lock_id` because I renamed `Epoch` to `AcquiredLockId`